src/testdrive/src/action/kafka/ingest.rs

```diff
@@ -26,33 +26,77 @@ pub struct IngestAction {
     topic_prefix: String,
     partition: i32,
     format: Format,
+    key_format: Option<Format>,
     timestamp: Option<i64>,
     publish: bool,
     rows: Vec<String>,
 }
 
 enum Format {
+    Avro { schema: String },
+    Proto { message: String },
+    Bytes,
+}
+
+enum Encoder {
     Avro {
-        key_schema: Option<String>,
-        value_schema: String,
+        schema: Schema,
+        schema_id: i32,
     },
     Proto {
-        message: String,
+        parser: &'static dyn Fn(&str) -> Result<protobuf::DynMessage, failure::Error>,
+        validator: &'static dyn Fn(&[u8]) -> Result<Box<dyn fmt::Debug>, failure::Error>,
     },
     Bytes,
 }
 
+impl Encoder {
+    fn decode_into_buf(&self, row: &str, buf: &mut Vec<u8>) -> Result<(), String> {
+        match self {
+            Encoder::Avro { schema, schema_id } => {
+                let val = crate::format::avro::json_to_avro(
+                    &serde_json::from_str(row)
+                        .map_err(|e| format!("parsing avro datum: {}", e.to_string()))?,
+                    schema.top_node(),
+                )?;
+                // The first byte is a magic byte (0) that indicates the Confluent
+                // serialization format version, and the next four bytes are a
+                // 32-bit schema ID.
+                //
+                // https://docs.confluent.io/current/schema-registry/docs/serializer-formatter.html#wire-format
+                buf.write_u8(0).unwrap();
+                buf.write_i32::<NetworkEndian>(*schema_id).unwrap();
+                buf.extend(avro::to_avro_datum(&schema, val).map_err(|e| e.to_string())?);
+            }
+            Encoder::Proto { parser, validator } => {
+                let msg =
+                    parser(row).map_err(|e| format!("converting row to type {} -> {}", row, e))?;
+                *buf = msg
+                    .write_to_bytes()
+                    .map_err(|e| format!("writing protobuf message for {}: {}", row, e))?;
+                // There are a variety of `write_*` methods on `Message` that don't
+                // seem to automatically do the right thing. This should always
+                // succeed, otherwise there is no chance for the server.
+                let _parsed = validator(&buf)
+                    .map_err(|e| format!("error validating proto row={}\nerror={}", row, e))?;
+            }
+            Encoder::Bytes => {
+                let json_string: serde_json::Value = serde_json::from_str(row)
+                    .map_err(|e| format!("parsing json string datum: {}", e.to_string()))?;
+                *buf = json_string.as_str().unwrap().as_bytes().to_vec();
```

Comment:
I still don't understand this. Why are you validating that these are JSON? `bytes` is not meant to enforce any particular format on its values. Shouldn't this just be:

```
*buf = row.as_bytes();
```